agent/src/com/thoughtworks/go/agent/service/AgentWebsocketService.java

```diff
@@ -0,0 +1,212 @@
+        sslContextFactory.setTrustStorePassword(SslInfrastructureService.AGENT_STORE_PASSWORD);
+        sslContextFactory.setWantClientAuth(true);
+        if (client == null || client.isStopped()) {
+            SystemEnvironment environment = new SystemEnvironment();
+            client = new WebSocketClient(sslContextFactory);
+            client.setMaxIdleTimeout(environment.getWebsocketMaxIdleTime());
+            Message.setupPolicy(client.getPolicy(), environment);
+            client.start();
+        }
+        if (session != null) {
+            session.close();
+        }
+        LOGGER.info("Connecting to websocket endpoint: " + urlService.getAgentRemoteWebSocketUrl());
+        session = client.connect(this, new URI(urlService.getAgentRemoteWebSocketUrl()), new ClientUpgradeRequest()).get();
+    }
+
+    public synchronized void stop() {
+        if (isRunning()) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("close " + sessionName());
+            }
+            session.close();
+            session = null;
+        }
+    }
+
+    public synchronized boolean isRunning() {
+        return session != null && session.isOpen();
+    }
+
+    public void send(Message message) {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(sessionName() + " send message: " + message);
+        }
+        try {
+            this.session.getRemote().sendBytes(ByteBuffer.wrap(Message.encode(message)));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @OnWebSocketConnect
+    public void onConnect(Session session) {
+        LOGGER.info(session + " connected.");
+    }
+
+    @OnWebSocketMessage
+    public void onMessage(InputStream raw) {
+        final Message msg = Message.decode(raw);
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(sessionName() + " message: " + msg);
+        }
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    AgentWebsocketService.this.controller.process(msg);
+                } catch (InterruptedException e) {
+                    LOGGER.error("Process message[" + msg + "] is interruptted.", e);
+                } catch (RuntimeException e) {
+                    LOGGER.error("Unexpected error while processing message[" + msg + "]: " + e.getMessage(), e);
+                }
+            }
+        });
+    }
+
+    @OnWebSocketClose
+    public void onClose(int closeCode, String closeReason) {
+        LOGGER.info(sessionName() + " closed.");
+    }
+
+    @OnWebSocketError
+    public void onError(Throwable error) {
+        LOGGER.error(sessionName() + " error", error);
+        if (error instanceof MessageTooLargeException) {
+            LOGGER.error("You can set Java system property 'go.websocket.max.text.message.size' to increase limit");
```

Comment:
The property name does not match the one in SystemEnvironment, may be you could use `SystemEnvironment.GO_WEBSOCKET_MAX_MESSAGE_SIZE.propertyName()`